pkgs/build-support/setup-hooks/patch-shebangs.sh

```diff
@@ -29,75 +32,76 @@ patchShebangs() {
-                sed -i -e "1 s|.*|#\!$escapedInterpreterLine|" "$f"
-                touch -r "$timestamp" "$f"
-                rm "$timestamp"
-            fi
-        fi
-    done < <(find "$dir" -type f -perm -0100 -print0)
-
-    stopNest
+    for dir in "$@"
+    do
+      header "patching script interpreter paths in $dir"
+      local f
+      local oldPath
+      local newPath
+      local arg0
+      local args
+      local oldInterpreterLine
+      local newInterpreterLine
+
+      [ -e "$dir" ] || return 0
+
+      local f
+      while IFS= read -r -d $'\0' f; do
+          isScript "$f" || continue
+
+          oldInterpreterLine=$(head -1 "$f" | tail -c+3)
+          read -r oldPath arg0 args <<< "$oldInterpreterLine"
+
+          if [ -z "$pathName" ]; then
+              if [ -n "$strictDeps" ] && [[ "$f" = "$NIX_STORE"* ]]; then
+                  pathName=HOST_PATH
+              else
+                  pathName=PATH
+              fi
+          fi
+
+          if $(echo "$oldPath" | grep -q "/bin/env$"); then
+              # Check for unsupported 'env' functionality:
+              # - options: something starting with a '-'
+              # - environment variables: foo=bar
+              if $(echo "$arg0" | grep -q -- "^-.*\|.*=.*"); then
+                  echo "$f: unsupported interpreter directive \"$oldInterpreterLine\" (set dontPatchShebangs=1 and handle shebang patching yourself)"
+                  exit 1
+              fi
+
+              newPath="$(PATH="${!pathName}" command -v "$arg0" || true)"
+          else
+              if [ "$oldPath" = "" ]; then
+                  # If no interpreter is specified linux will use /bin/sh. Set
+                  # oldpath="/bin/sh" so that we get /nix/store/.../sh.
+                  oldPath="/bin/sh"
+              fi
+
+              newPath="$(PATH="${!pathName}" command -v "$(basename "$oldPath")" || true)"
+
+              args="$arg0 $args"
+          fi
+
+          # Strip trailing whitespace introduced when no arguments are present
+          newInterpreterLine="$(echo "$newPath $args" | sed 's/[[:space:]]*$//')"
+
+          if [ -n "$oldPath" -a "${oldPath:0:${#NIX_STORE}}" != "$NIX_STORE" ]; then
+              if [ -n "$newPath" -a "$newPath" != "$oldPath" ]; then
+                  echo "$f: interpreter directive changed from \"$oldInterpreterLine\" to \"$newInterpreterLine\""
+                  # escape the escape chars so that sed doesn't interpret them
+                  escapedInterpreterLine=$(echo "$newInterpreterLine" | sed 's|\\|\\\\|g')
+                  # Preserve times, see: https://github.com/NixOS/nixpkgs/pull/33281
+                  timestamp=$(mktemp)
+                  touch -r "$f" "$timestamp"
+                  sed -i -e "1 s|.*|#\!$escapedInterpreterLine|" "$f"
+                  touch -r "$timestamp" "$f"
+                  rm "$timestamp"
+              fi
+          fi
+      done < <(find "$dir" -type f -perm -0100 -print0)
```

Comment:
Could you just put ```find "$@"``` here? I think that could be more efficient

Reply:
Oh, that would remove the need for my loop, right?

Reply:
Yeah would just have to remove the ```[ -e "$dir" ] || return 0``` part as well.

Reply:
Thanks! I tried the function out on its own and it seems to be doing the right things.